src/client/Microsoft.Identity.Client/Internal/Requests/CredentialBasedMsiAuthRequest.cs

```diff
@@ -0,0 +1,327 @@
+        /// </summary>
+        /// <param name="keyMaterial"></param>
+        /// <param name="httpManager"></param>
+        /// <param name="credentialResponse"></param>
+        /// <returns></returns>
+        private OAuth2Client CreateClientRequest(
+            IKeyMaterialManager keyMaterial,
+            IHttpManager httpManager,
+            CredentialResponse credentialResponse)
+        {
+            // Initialize an OAuth2 client with logger, HTTP manager, and binding certificate.
+            var client = new OAuth2Client(
+                AuthenticationRequestParameters.RequestContext.Logger,
+                httpManager,
+                keyMaterial.BindingCertificate);
+
+            // Convert overridden scopes to a single string.
+            string scopes = GetOverriddenScopes(AuthenticationRequestParameters.Scope).AsSingleString();
+
+            // Add required parameters for client credentials grant request.
+            client.AddBodyParameter(OAuth2Parameter.GrantType, OAuth2GrantType.ClientCredentials);
+            client.AddBodyParameter(OAuth2Parameter.Scope, scopes);
+            client.AddBodyParameter(OAuth2Parameter.ClientId, credentialResponse.ClientId);
+            client.AddBodyParameter(OAuth2Parameter.ClientAssertion, credentialResponse.Credential);
+            client.AddBodyParameter(OAuth2Parameter.ClientAssertionType, OAuth2AssertionType.JwtBearer);
+
+            // Add optional claims and client capabilities parameter if provided.
+            if (!string.IsNullOrWhiteSpace(AuthenticationRequestParameters.ClaimsAndClientCapabilities))
+            {
+                client.AddBodyParameter(OAuth2Parameter.Claims, AuthenticationRequestParameters.ClaimsAndClientCapabilities);
+            }
+
+            // Return the configured OAuth2 client.
+            return client;
+        }
+
+        // Check if CredentialKeyType is set to a valid value for Managed Identity.
+        private static bool IsCredentialKeyAvailable(
+            RequestContext requestContext,
+            AcquireTokenForManagedIdentityParameters managedIdentityParameters,
+            out Uri credentialEndpointUri)
+        {
+            credentialEndpointUri = null;
+
+            CryptoKeyType credentialKeyType = requestContext.ServiceBundle.Config.ManagedIdentityCredentialKeyType;
+
+            // CredentialKeyType will be None, if no keys are provisioned.
+            if (credentialKeyType == CryptoKeyType.None)
+            {
+                // Check to see if CAE or POP APIs are used.
+                bool isCaeOrPopRequested = requestContext.ServiceBundle.Config.ClientCapabilities?.Any() == true ||
+                                           requestContext.ServiceBundle.Config.ManagedIdentityPopSupported;
+
+                if (isCaeOrPopRequested)
+                {
+                    // Determine the specific unsupported feature to provide a clearer error message
```

Comment:
I think the initial check for isCaeOrPopRequested can be removed, along with this if/else block without changing the behavior.